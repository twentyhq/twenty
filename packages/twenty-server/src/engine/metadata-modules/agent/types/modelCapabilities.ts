export type ModelCapabilities = {
  webSearch?: boolean;
};
