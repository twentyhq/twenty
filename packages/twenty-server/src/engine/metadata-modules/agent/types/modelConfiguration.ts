export type ModelConfiguration = {
  webSearch?: {
    enabled: boolean;
    configuration: object;
  };
  twitterSearch?: {
    enabled: boolean;
    configuration: object;
  };
};
