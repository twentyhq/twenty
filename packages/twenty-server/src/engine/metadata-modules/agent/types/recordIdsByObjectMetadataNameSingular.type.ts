export type RecordIdsByObjectMetadataNameSingularType = Array<{
  objectMetadataNameSingular: string;
  recordIds: string[];
}>;
