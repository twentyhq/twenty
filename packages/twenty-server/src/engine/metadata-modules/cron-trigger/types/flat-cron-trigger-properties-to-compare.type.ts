import { type FlatCronTrigger } from 'src/engine/metadata-modules/cron-trigger/types/flat-cron-trigger.type';

export type FlatCronTriggerPropertiesToCompare = keyof Pick<
  FlatCronTrigger,
  'settings'
>;
