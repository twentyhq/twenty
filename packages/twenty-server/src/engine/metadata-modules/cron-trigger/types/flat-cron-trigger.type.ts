import { type WorkspaceEntity } from 'src/engine/core-modules/workspace/workspace.entity';
import { type CronTrigger } from 'src/engine/metadata-modules/cron-trigger/entities/cron-trigger.entity';
import { type FlatEntityFrom } from 'src/engine/metadata-modules/flat-entity/types/flat-entity.type';
import { type ServerlessFunctionEntity } from 'src/engine/metadata-modules/serverless-function/serverless-function.entity';
import { type ExtractRecordTypeOrmRelationProperties } from 'src/engine/workspace-manager/workspace-migration-v2/types/extract-record-typeorm-relation-properties.type';

export type CronTriggerEntityRelationProperties =
  ExtractRecordTypeOrmRelationProperties<
    CronTrigger,
    ServerlessFunctionEntity | WorkspaceEntity
  >;

export type FlatCronTrigger = FlatEntityFrom<
  CronTrigger,
  CronTriggerEntityRelationProperties
>;
