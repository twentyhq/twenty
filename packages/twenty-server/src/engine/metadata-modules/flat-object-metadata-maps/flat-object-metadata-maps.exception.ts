import {
  appendCommonExceptionCode,
  CustomException,
} from 'src/utils/custom-exception';

export class FlatObjectMetadataMapsException extends CustomException<
  keyof typeof FlatObjectMetadataMapsExceptionCode
> {}

export const FlatObjectMetadataMapsExceptionCode = appendCommonExceptionCode({
  FIELD_METADATA_NOT_FOUND: 'FIELD_METADATA_NOT_FOUND',
  OBJECT_METADATA_NOT_FOUND: 'OBJECT_METADATA_NOT_FOUND',
  FIELD_METADATA_ALREADY_EXISTS: 'FIELD_METADATA_ALREADY_EXISTS',
  OBJECT_METADATA_ALREADY_EXISTS: 'OBJECT_METADATA_ALREADY_EXISTS',
} as const);
