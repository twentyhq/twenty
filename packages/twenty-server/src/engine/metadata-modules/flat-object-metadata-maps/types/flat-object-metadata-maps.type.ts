import { type FlatObjectMetadataWithFlatFieldMaps } from 'src/engine/metadata-modules/flat-object-metadata-maps/types/flat-object-metadata-with-flat-field-metadata-maps.type';

export type FlatObjectMetadataMaps = {
  byId: Partial<Record<string, FlatObjectMetadataWithFlatFieldMaps>>;
  idByNameSingular: Partial<Record<string, string>>;
};
