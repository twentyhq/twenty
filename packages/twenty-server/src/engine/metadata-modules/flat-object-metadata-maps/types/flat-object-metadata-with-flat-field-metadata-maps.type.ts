import { type FlatFieldMetadata } from 'src/engine/metadata-modules/flat-field-metadata/types/flat-field-metadata.type';
import { type FlatObjectMetadata } from 'src/engine/metadata-modules/flat-object-metadata/types/flat-object-metadata.type';

export type FlatFieldMetadataMaps = {
  fieldsById: Partial<Record<string, FlatFieldMetadata>>;
  fieldIdByJoinColumnName: Partial<Record<string, string>>;
  fieldIdByName: Partial<Record<string, string>>;
};

export type FlatObjectMetadataWithFlatFieldMaps = FlatObjectMetadata &
  FlatFieldMetadataMaps;
