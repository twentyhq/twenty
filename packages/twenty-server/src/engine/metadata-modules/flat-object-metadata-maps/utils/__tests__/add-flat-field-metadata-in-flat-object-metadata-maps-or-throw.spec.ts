import {
  type EachTestingContext,
  eachTestingContextFilter,
} from 'twenty-shared/testing';
import { FieldMetadataType } from 'twenty-shared/types';

import { RelationType } from 'src/engine/metadata-modules/field-metadata/interfaces/relation-type.interface';

import { getFlatFieldMetadataMock } from 'src/engine/metadata-modules/flat-field-metadata/__mocks__/get-flat-field-metadata.mock';
import { PET_FLAT_FIELDS_MOCK } from 'src/engine/metadata-modules/flat-field-metadata/__mocks__/pet-flat-fields.mock';
import { expectFlatFieldMetadataToBeInFlatObjectMetadataMaps } from 'src/engine/metadata-modules/flat-object-metadata-maps/__tests__/utils/expect-flat-field-metadata-to-be-in-flat-object-metadata-maps.util.test';
import { FLAT_OBJECT_METADATA_MAPS_MOCKS } from 'src/engine/metadata-modules/flat-object-metadata-maps/mocks/flat-object-metadata-maps.mock';
import {
  type AddFlatFieldMetadataInFlatObjectMetadataMapsOrThrowArgs,
  addFlatFieldMetadataInFlatObjectMetadataMapsOrThrow,
} from 'src/engine/metadata-modules/flat-object-metadata-maps/utils/add-flat-field-metadata-in-flat-object-metadata-maps-or-throw.util';
import { PET_FLAT_OBJECT_MOCK } from 'src/engine/metadata-modules/flat-object-metadata/__mocks__/pet-flat-object.mock';
import { RelationOnDeleteAction } from 'src/engine/metadata-modules/relation-metadata/relation-on-delete-action.type';

type AddFlatFieldMetadataInFlatObjectMetadataMapsOrThrowArgsTestCase = {
  input: AddFlatFieldMetadataInFlatObjectMetadataMapsOrThrowArgs;
  shouldThrow?: true;
};

describe('addFlatFieldMetadataInFlatObjectMetadataMapsOrThrow', () => {
  const testCases: EachTestingContext<AddFlatFieldMetadataInFlatObjectMetadataMapsOrThrowArgsTestCase>[] =
    [
      {
        title:
          'should throw when field metadata parent object metadata does not exist',
        context: {
          shouldThrow: true,
          input: {
            flatObjectMetadataMaps: FLAT_OBJECT_METADATA_MAPS_MOCKS,
            flatFieldMetadata: getFlatFieldMetadataMock({
              objectMetadataId: 'non-existent',
              type: FieldMetadataType.TEXT,
              uniqueIdentifier: 'unique-id-1',
              id: 'unique-id-1',
            }),
          },
        },
      },
      {
        title:
          'should throw when field metadata to add already exist in object',
        context: {
          shouldThrow: true,
          input: {
            flatObjectMetadataMaps: FLAT_OBJECT_METADATA_MAPS_MOCKS,
            flatFieldMetadata: PET_FLAT_FIELDS_MOCK.species,
          },
        },
      },
      {
        title: 'should add a new relation field to pet object',
        context: {
          input: {
            flatObjectMetadataMaps: FLAT_OBJECT_METADATA_MAPS_MOCKS,
            flatFieldMetadata: getFlatFieldMetadataMock({
              objectMetadataId: PET_FLAT_OBJECT_MOCK.id,
              type: FieldMetadataType.RELATION,
              uniqueIdentifier: 'field-metadata-id',
              id: 'field-metadata-id',
              settings: {
                relationType: RelationType.MANY_TO_ONE,
                joinColumnName: 'whateverId',
                onDelete: RelationOnDeleteAction.CASCADE,
              },
              relationTargetObjectMetadataId: PET_FLAT_OBJECT_MOCK.id,
              relationTargetFieldMetadataId: 'does-not-matter',
            }),
          },
        },
      },
    ];

  const filteredTestCases = eachTestingContextFilter(testCases);

  it.each(filteredTestCases)(
    '$title',
    ({
      context: {
        input: { flatFieldMetadata, flatObjectMetadataMaps },
        shouldThrow = false,
      },
    }) => {
      if (shouldThrow) {
        expect(() =>
          addFlatFieldMetadataInFlatObjectMetadataMapsOrThrow({
            flatFieldMetadata,
            flatObjectMetadataMaps,
          }),
        ).toThrowErrorMatchingSnapshot();
      } else {
        const updatedFlatObjectMetadataMaps =
          addFlatFieldMetadataInFlatObjectMetadataMapsOrThrow({
            flatFieldMetadata,
            flatObjectMetadataMaps,
          });

        expectFlatFieldMetadataToBeInFlatObjectMetadataMaps({
          flatFieldMetadata,
          flatObjectMetadataMaps: updatedFlatObjectMetadataMaps,
        });
      }
    },
  );
});
