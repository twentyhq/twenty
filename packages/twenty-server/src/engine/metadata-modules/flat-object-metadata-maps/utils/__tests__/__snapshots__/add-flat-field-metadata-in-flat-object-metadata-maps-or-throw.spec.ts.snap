// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addFlatFieldMetadataInFlatObjectMetadataMapsOrThrow should throw when field metadata parent object metadata does not exist 1`] = `"addFlatFieldMetadataInFlatObjectMetadataMapsOrThrow field parent object metadata not found"`;

exports[`addFlatFieldMetadataInFlatObjectMetadataMapsOrThrow should throw when field metadata to add already exist in object 1`] = `"addFlatFieldMetadataToFlatObjectMetadataWithFlatFieldMaps added flatFieldMetadata already exists"`;
